File: Balanced Binary Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(checkBalanced(root)==-1)
            return false;
        else
            return true;
    }
    
    int checkBalanced(TreeNode *root)
    {
        if(root==NULL)
            return 0;
        
        int left = checkBalanced(root->left);
        
        if(left==-1)
            return -1;
        else
        {
            int right = checkBalanced(root->right);
            
            if(right==-1)
                return -1;
            else
            {
                int val = left-right;
                if(val<-1||val>1)
                    return -1;
                else
                {
                    val = left>right?left:right;
                    return val+1;
                }
                    
            }
                
        }
      
    }
};

File: Balanced Binary Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        boolean ret = true;
        if(root != null)
        {
            ret = (isBalancedBasedOnDepth(root) == -1)? false: true;
        }
        return ret;
    }
    
    private int isBalancedBasedOnDepth(TreeNode root)
    {
        if(root == null)
        {
            return 0;
        }
        
        int leftDepth = isBalancedBasedOnDepth(root.left);
        int rightDepth = (leftDepth == -1)? -1:isBalancedBasedOnDepth(root.right);
        
        if(leftDepth < 0|| rightDepth < 0 ||Math.abs(leftDepth-rightDepth)>1)
        {
            return -1;
        }
        else
        {
            return Math.max(leftDepth, rightDepth) + 1;
        }
    }
}